src/hotspot/share/services/threadService.hpp

```diff
@@ -102,6 +106,12 @@ class ThreadService : public AllStatic {
   static jlong get_live_thread_count()        { return _atomic_threads_count; }
   static jlong get_daemon_thread_count()      { return _atomic_daemon_threads_count; }
 
+  static jlong exited_allocated_bytes()       { return _exited_allocated_bytes; }
+  static void incr_exited_allocated_bytes(jlong size) {
+    // No need for atomicity, method is called under the Threads_lock
+    _exited_allocated_bytes += size;
+  }
```

Comment:
As there is a lock-free read however you need to use Atomic::load and Atomic::store - epsecially for 32-bit systems.

Reply:
Changed. Not sure atomic load is needed.

Reply:
Even though this version passed the pre-submit tests, which include the new versions of ThreadAllocatedBytes.java, running it more often on a host with many hardware threads resulted in failures where getTotalAllocatedBytes did not monotonically increase. I believe what happened was that without an acquire/release, the hardware can float earlier stores to _exited_allocated_bytes above later ones. Adding acquire/release fixed the issue.

Reply:
That should not be possible if the update is always under the lock. acquire/release control what other memory updates you see when you see "this" update - they have no affect on the update itself.